### ISSUE TITLE
Adapter experiments

### ISSUE THREAD

**Issue by Vectorrent** [0-5788719150923125/praxis#28]
I want to experiment with something like a "swarm of LoRA adapters," connected to the base model. This might be something like a [LoRA Mixture-of-Experts](https://huggingface.co/docs/peft/en/package_reference/xlora), or it might be something a little more interesting, such as:

Imagine if every expert/layer in the model had "access" to a pool of LoRA modules, and if the model could use sparse activations to intelligently-select certain experts to use, at certain steps of computation? Rather than activating all experts, all at once - one could activate the top-k experts, as needed.

However, I am not interested in pretraining a base model, then pretraining a bunch of adapters after. Instead, I would integrate the LoRA training into the pretraining itself... somehow. 

The current idea involves building a Pytorch module that can be used to manage the entire lifecycle of an entire pool of LoRA adapters:

- initializing LoRA adapters
- merging adapters
- resetting adapters
- switching/routing through adapters
- enabling/disabling gradient computation for targeted experts
- etc...

Mainly, the main idea I'm having involves the integration of training code, **directly into the Pytorch Module itself**. We would not use the main training loop for updating LoRA modules; instead, we would train each module independently (or perhaps collaboratively). There would be a strong correlation between the base model and adapters, but the actual management of a training loop, optimizer, and other components - all of that would be built-in to the module itself. And thus, during training, a person would have 1 main model in-training, and a million tiny experts (or maybe huge ones) switching on/off, constantly.

**Comment by Vectorrent** [comment]
It's hard to imagine how this would ever work, given the fact that Praxis is about model research and development - and models are NEVER frozen, as it currently stands. Models are constantly churning. The one place LoRAs might work would be in remote-peering scenarios, where remote expert models could be "mixed" with a locally-training model. But this is way, way into the future - and not worth tracking on Github. It's too abstract for now.
